File: cpp/src/remoting/protocol/extra_info.cpp
#include "rocketmq/remoting/protocol/extra_info.h"

#include <stdexcept>

#include "rocketmq/common/mix_all.h"

namespace rocketmq {
namespace extra_info {

namespace {

constexpr char kRetrySepV1 = '_';
constexpr char kRetrySepV2 = '+';

// 复刻 Java String.split(sep) 的"丢弃末尾空串"行为。
std::vector<std::string> splitAndDropTrailing(const std::string& value, char sep) {
    std::vector<std::string> parts;
    std::size_t start = 0;
    while (true) {
        std::size_t pos = value.find(sep, start);
        if (pos == std::string::npos) {
            parts.push_back(value.substr(start));
            break;
        }
        parts.push_back(value.substr(start, pos - start));
        start = pos + 1;
    }
    while (!parts.empty() && parts.back().empty()) {
        parts.pop_back();
    }
    return parts;
}

void requireSize(const std::vector<std::string>& segments, std::size_t need, const char* what) {
    if (segments.size() < need) {
        throw std::invalid_argument(std::string(what) + " fail, extraInfoStrs length "
                                    + std::to_string(segments.size()));
    }
}

int64_t toInt64(const std::string& s) {
    return static_cast<int64_t>(std::stoll(s));
}

int32_t toInt32(const std::string& s) {
    return static_cast<int32_t>(std::stol(s));
}

template <typename Value, typename Convert>
std::optional<std::map<std::string, Value>> parseThreeFieldInfo(
    const std::string& raw, const char* what, Convert convert) {
    if (raw.empty()) {
        return std::nullopt;
    }
    std::map<std::string, Value> out;
    std::vector<std::string> segments;
    if (raw.find(kQueueSeparator) == std::string::npos) {
        segments.push_back(raw);
    } else {
        segments = splitAndDropTrailing(raw, kQueueSeparator);
    }
    for (const std::string& one : segments) {
        std::vector<std::string> parts = splitAndDropTrailing(one, ' ');
        if (parts.size() != 3) {
            throw std::invalid_argument(std::string("parse ") + what + " error, " + raw);
        }
        std::string key = parts[0] + "@" + parts[1];
        if (out.find(key) != out.end()) {
            throw std::invalid_argument(std::string("parse ") + what + " error, duplicate, " + raw);
        }
        out[key] = convert(parts[2]);
    }
    return out;
}

std::string join(const std::vector<std::string>& parts, const char* sep) {
    std::string out;
    for (std::size_t i = 0; i < parts.size(); ++i) {
        if (i > 0) {
            out += sep;
        }
        out += parts[i];
    }
    return out;
}

}  // namespace

std::string buildPopRetryTopicV1(const std::string& topic, const std::string& cid) {
    return std::string(MixAll::RETRY_GROUP_TOPIC_PREFIX) + cid + kRetrySepV1 + topic;
}

std::string buildPopRetryTopicV2(const std::string& topic, const std::string& cid) {
    return std::string(MixAll::RETRY_GROUP_TOPIC_PREFIX) + cid + kRetrySepV2 + topic;
}

std::string buildPopRetryTopic(const std::string& topic, const std::string& cid,
                               bool enableRetryV2) {
    if (enableRetryV2) {
        return buildPopRetryTopicV2(topic, cid);
    }
    return buildPopRetryTopicV1(topic, cid);
}

bool isPopRetryTopicV2(const std::string& retryTopic) {
    if (retryTopic.empty()) {
        return false;
    }
    return retryTopic.rfind(MixAll::RETRY_GROUP_TOPIC_PREFIX, 0) == 0
           && retryTopic.find(kRetrySepV2) != std::string::npos;
}

std::vector<std::string> split(const std::string& extraInfo) {
    return splitAndDropTrailing(extraInfo, ' ');
}

int64_t getCkQueueOffset(const std::vector<std::string>& segments) {
    requireSize(segments, 1, "getCkQueueOffset");
    return toInt64(segments[0]);
}

int64_t getPopTime(const std::vector<std::string>& segments) {
    requireSize(segments, 2, "getPopTime");
    return toInt64(segments[1]);
}

int64_t getInvisibleTime(const std::vector<std::string>& segments) {
    requireSize(segments, 3, "getInvisibleTime");
    return toInt64(segments[2]);
}

int32_t getReviveQid(const std::vector<std::string>& segments) {
    requireSize(segments, 4, "getReviveQid");
    return toInt32(segments[3]);
}

std::string getRetry(const std::vector<std::string>& segments) {
    requireSize(segments, 5, "getRetry");
    return segments[4];
}

std::string getBrokerName(const std::vector<std::string>& segments) {
    requireSize(segments, 6, "getBrokerName");
    return segments[5];
}

int32_t getQueueId(const std::vector<std::string>& segments) {
    requireSize(segments, 7, "getQueueId");
    return toInt32(segments[6]);
}

int64_t getQueueOffset(const std::vector<std::string>& segments) {
    requireSize(segments, 8, "getQueueOffset");
    return toInt64(segments[7]);
}

std::string retryOfTopic(const std::string& topic) {
    if (isPopRetryTopicV2(topic)) {
        return kRetryTopicV2;
    }
    if (topic.rfind(MixAll::RETRY_GROUP_TOPIC_PREFIX, 0) == 0) {
        return kRetryTopic;
    }
    return kNormalTopic;
}

std::string buildExtraInfo(int64_t ckQueueOffset, int64_t popTime, int64_t invisibleTime,
                           int32_t reviveQid, const std::string& topic,
                           const std::string& brokerName, int32_t queueId) {
    std::vector<std::string> parts = {
        std::to_string(ckQueueOffset),
        std::to_string(popTime),
        std::to_string(invisibleTime),
        std::to_string(reviveQid),
        retryOfTopic(topic),
        brokerName,
        std::to_string(queueId),
    };
    return join(parts, kKeySeparator);
}

std::string buildExtraInfo(int64_t ckQueueOffset, int64_t popTime, int64_t invisibleTime,
                           int32_t reviveQid, const std::string& topic,
                           const std::string& brokerName, int32_t queueId,
                           int64_t msgQueueOffset) {
    return buildExtraInfo(ckQueueOffset, popTime, invisibleTime, reviveQid, topic, brokerName,
                          queueId)
           + kKeySeparator + std::to_string(msgQueueOffset);
}

std::optional<std::map<std::string, int64_t>> parseStartOffsetInfo(const std::string& s) {
    return parseThreeFieldInfo<int64_t>(s, "startOffsetInfo",
                                        [](const std::string& v) { return toInt64(v); });
}

std::optional<std::map<std::string, std::vector<int64_t>>> parseMsgOffsetInfo(
    const std::string& s) {
    return parseThreeFieldInfo<std::vector<int64_t>>(
        s, "msgOffsetInfo", [](const std::string& v) {
            std::vector<int64_t> offsets;
            for (const std::string& one : splitAndDropTrailing(v, kOffsetSeparator)) {
                offsets.push_back(toInt64(one));
            }
            return offsets;
        });
}

std::optional<std::map<std::string, int32_t>> parseOrderCountInfo(const std::string& s) {
    return parseThreeFieldInfo<int32_t>(s, "orderCountInfo",
                                        [](const std::string& v) { return toInt32(v); });
}

std::string getStartOffsetInfoMapKey(const std::string& topic, int64_t key) {
    return retryOfTopic(topic) + "@" + std::to_string(key);
}

std::string getQueueOffsetKeyValueKey(int64_t queueId, int64_t queueOffset) {
    return std::string("qo") + std::to_string(queueId) + "%" + std::to_string(queueOffset);
}

std::string getQueueOffsetMapKey(const std::string& topic, int64_t queueId,
                                 int64_t queueOffset) {
    return retryOfTopic(topic) + "@" + getQueueOffsetKeyValueKey(queueId, queueOffset);
}

bool isOrder(const std::vector<std::string>& segments) {
    return getReviveQid(segments) == kPopOrderReviveQueue;
}

std::string getRealTopic(const std::string& topic, const std::string& cid,
                         const std::string& retry) {
    if (retry == kNormalTopic) {
        return topic;
    }
    if (retry == kRetryTopic) {
        return buildPopRetryTopicV1(topic, cid);
    }
    if (retry == kRetryTopicV2) {
        return buildPopRetryTopicV2(topic, cid);
    }
    throw std::invalid_argument("getRetry fail, format is wrong");
}

}  // namespace extra_info
}  // namespace rocketmq

File: python/rocketmq/remoting/protocol/extra_info.py
"""POP 模式的 extraInfo（俗称 CK 串）编解码。

逐条移植 Java ``org.apache.rocketmq.remoting.protocol.header.ExtraInfoUtil``。

CK 串是 POP 模式的核心凭据：broker 在 POP 响应里**不**给普通 topic 的消息写
``POP_CK`` 属性（只在 retry topic 重编码路径写），客户端必须自己用响应头的
``startOffsetInfo`` / ``msgOffsetInfo`` 反构出这个串，再作为 ACK /
CHANGE_MESSAGE_INVISIBLETIME 的 ``extraInfo`` 回传。格式是 8 段、**空格**分隔：

``ckQueueOffset popTime invisibleTime reviveQid retryFlag brokerName queueId [msgQueueOffset]``

注意分隔符是空格而不是逗号 —— 用错会导致 broker 静默解析失败。
"""

from typing import Dict, List, Optional

from rocketmq.common.mix_all import MixAll

#: 段内分隔符；**必须**是空格，对应 Java ``MessageConst.KEY_SEPARATOR``
KEY_SEPARATOR = " "
#: 队列之间的分隔符
QUEUE_SEPARATOR = ";"
#: ``msgOffsetInfo`` 里同一队列多条 offset 的分隔符
OFFSET_SEPARATOR = ","

NORMAL_TOPIC = "0"
RETRY_TOPIC = "1"
RETRY_TOPIC_V2 = "2"
QUEUE_OFFSET = "qo"

#: 顺序消费用的固定 revive 队列号，对应 Java ``KeyBuilder.POP_ORDER_REVIVE_QUEUE``
POP_ORDER_REVIVE_QUEUE = 999

_POP_RETRY_SEPARATOR_V1 = "_"
_POP_RETRY_SEPARATOR_V2 = "+"


def build_pop_retry_topic_v1(topic: str, cid: str) -> str:
    """``%RETRY%<cid>_<topic>``（Java ``KeyBuilder.buildPopRetryTopicV1``）。"""
    return MixAll.RETRY_GROUP_TOPIC_PREFIX + cid + _POP_RETRY_SEPARATOR_V1 + topic


def build_pop_retry_topic_v2(topic: str, cid: str) -> str:
    """``%RETRY%<cid>+<topic>``（Java ``KeyBuilder.buildPopRetryTopicV2``）。"""
    return MixAll.RETRY_GROUP_TOPIC_PREFIX + cid + _POP_RETRY_SEPARATOR_V2 + topic


def build_pop_retry_topic(topic: str, cid: str, enable_retry_v2: bool = False) -> str:
    """Java ``buildPopRetryTopic``：``enableRetryTopicV2`` 关（默认）走 V1。"""
    if enable_retry_v2:
        return build_pop_retry_topic_v2(topic, cid)
    return build_pop_retry_topic_v1(topic, cid)


def is_pop_retry_topic_v2(retry_topic: Optional[str]) -> bool:
    """Java ``KeyBuilder.isPopRetryTopicV2``：``%RETRY%`` 前缀且含 ``+``。"""
    if not retry_topic:
        return False
    return retry_topic.startswith(MixAll.RETRY_GROUP_TOPIC_PREFIX) and (
        _POP_RETRY_SEPARATOR_V2 in retry_topic
    )


def _split_drop_trailing(value: str, sep: str) -> List[str]:
    """模拟 Java ``String.split``：**丢弃末尾空串**（Python 的 ``split`` 会保留）。

    这个差异是真实的：Java ``"a b ".split(" ")`` 得到 ``["a","b"]``，
    而 Python ``"a b ".split(" ")`` 得到 ``["a","b",""]``。段数校验依赖它。
    """
    parts = value.split(sep)
    while parts and parts[-1] == "":
        parts.pop()
    return parts


def split(extra_info: Optional[str]) -> List[str]:
    """按空格切分 CK 串（Java ``ExtraInfoUtil.split``）。"""
    if extra_info is None:
        raise ValueError("split extraInfo is null")
    return _split_drop_trailing(extra_info, KEY_SEPARATOR)


def _require(segments: Optional[List[str]], need: int, what: str) -> None:
    if segments is None or len(segments) < need:
        raise ValueError(
            "%s fail, extraInfoStrs length %d" % (what, 0 if segments is None else len(segments))
        )


def get_ck_queue_offset(segments: Optional[List[str]]) -> int:
    _require(segments, 1, "getCkQueueOffset")
    return int(segments[0])


def get_pop_time(segments: Optional[List[str]]) -> int:
    _require(segments, 2, "getPopTime")
    return int(segments[1])


def get_invisible_time(segments: Optional[List[str]]) -> int:
    _require(segments, 3, "getInvisibleTime")
    return int(segments[2])


def get_revive_qid(segments: Optional[List[str]]) -> int:
    _require(segments, 4, "getReviveQid")
    return int(segments[3])


def get_retry(segments: Optional[List[str]]) -> str:
    _require(segments, 5, "getRetry")
    return segments[4]


def get_broker_name(segments: Optional[List[str]]) -> str:
    _require(segments, 6, "getBrokerName")
    return segments[5]


def get_queue_id(segments: Optional[List[str]]) -> int:
    _require(segments, 7, "getQueueId")
    return int(segments[6])


def get_queue_offset(segments: Optional[List[str]]) -> int:
    _require(segments, 8, "getQueueOffset")
    return int(segments[7])


def retry_of_topic(topic: str) -> str:
    """由 topic 形状判定 retryFlag（Java ``ExtraInfoUtil.getRetry(topic)``）。

    顺序很重要：先判 V2（含 ``+``），再判 ``%RETRY%`` 前缀（V1）。
    """
    if is_pop_retry_topic_v2(topic):
        return RETRY_TOPIC_V2
    if topic.startswith(MixAll.RETRY_GROUP_TOPIC_PREFIX):
        return RETRY_TOPIC
    return NORMAL_TOPIC


def build_extra_info(ck_queue_offset: int, pop_time: int, invisible_time: int,
                     revive_qid: int, topic: str, broker_name: str, queue_id: int,
                     msg_queue_offset: Optional[int] = None) -> str:
    """拼 CK 串。传 ``msg_queue_offset`` 得到 8 段，不传得到 7 段。

    对应 Java 的两个 ``buildExtraInfo`` 重载。ACK 场景用 8 段版本。
    """
    parts = [
        str(ck_queue_offset),
        str(pop_time),
        str(invisible_time),
        str(revive_qid),
        retry_of_topic(topic),
        str(broker_name),
        str(queue_id),
    ]
    if msg_queue_offset is not None:
        parts.append(str(msg_queue_offset))
    return KEY_SEPARATOR.join(parts)


def parse_start_offset_info(start_offset_info: Optional[str]) -> Optional[Dict[str, int]]:
    """解析 ``startOffsetInfo``，形如 ``"0 3 0;0 2 0"``。

    key 是 ``"<retryFlag>@<queueId>"``，value 是该队列本次弹出的起始 offset。
    """
    if not start_offset_info:
        return None
    out: Dict[str, int] = {}
    segments = (
        [start_offset_info]
        if QUEUE_SEPARATOR not in start_offset_info
        else _split_drop_trailing(start_offset_info, QUEUE_SEPARATOR)
    )
    for one in segments:
        parts = one.split(KEY_SEPARATOR)
        if len(parts) != 3:
            raise ValueError("parse startOffsetInfo error, " + str(start_offset_info))
        key = parts[0] + "@" + parts[1]
        if key in out:
            raise ValueError("parse startOffsetInfo error, duplicate, " + str(start_offset_info))
        out[key] = int(parts[2])
    return out


def parse_msg_offset_info(msg_offset_info: Optional[str]) -> Optional[Dict[str, List[int]]]:
    """解析 ``msgOffsetInfo``，形如 ``"0 3 0,1,2;0 2 0"``。

    key 同 ``parse_start_offset_info``，value 是该队列本次弹出的各条消息 offset 列表。
    """
    if not msg_offset_info:
        return None
    out: Dict[str, List[int]] = {}
    segments = (
        [msg_offset_info]
        if QUEUE_SEPARATOR not in msg_offset_info
        else _split_drop_trailing(msg_offset_info, QUEUE_SEPARATOR)
    )
    for one in segments:
        parts = one.split(KEY_SEPARATOR)
        if len(parts) != 3:
            raise ValueError("parse msgOffsetInfo error, " + str(msg_offset_info))
        key = parts[0] + "@" + parts[1]
        if key in out:
            raise ValueError("parse msgOffsetInfo error, duplicate, " + str(msg_offset_info))
        out[key] = [int(x) for x in _split_drop_trailing(parts[2], OFFSET_SEPARATOR)]
    return out


def parse_order_count_info(order_count_info: Optional[str]) -> Optional[Dict[str, int]]:
    """解析 ``orderCountInfo``（顺序消费用），第三段是计数。"""
    if not order_count_info:
        return None
    out: Dict[str, int] = {}
    segments = (
        [order_count_info]
        if QUEUE_SEPARATOR not in order_count_info
        else _split_drop_trailing(order_count_info, QUEUE_SEPARATOR)
    )
    for one in segments:
        parts = one.split(KEY_SEPARATOR)
        if len(parts) != 3:
            raise ValueError("parse orderCountInfo error, " + str(order_count_info))
        key = parts[0] + "@" + parts[1]
        if key in out:
            raise ValueError("parse orderCountInfo error, duplicate, " + str(order_count_info))
        out[key] = int(parts[2])
    return out


def get_start_offset_info_map_key(topic: str, key) -> str:
    return retry_of_topic(topic) + "@" + str(key)


def get_queue_offset_key_value_key(queue_id, queue_offset) -> str:
    return QUEUE_OFFSET + str(queue_id) + "%" + str(queue_offset)


def get_queue_offset_map_key(topic: str, queue_id, queue_offset) -> str:
    return retry_of_topic(topic) + "@" + get_queue_offset_key_value_key(queue_id, queue_offset)


def is_order(segments: List[str]) -> bool:
    """reviveQid 是 999 表示顺序消费（Java ``ExtraInfoUtil.isOrder``）。"""
    return get_revive_qid(segments) == POP_ORDER_REVIVE_QUEUE


def get_real_topic(topic: str, cid: str, retry: str) -> str:
    """由 retryFlag 还原真实 topic（Java ``ExtraInfoUtil.getRealTopic``）。"""
    if retry == NORMAL_TOPIC:
        return topic
    if retry == RETRY_TOPIC:
        return build_pop_retry_topic_v1(topic, cid)
    if retry == RETRY_TOPIC_V2:
        return build_pop_retry_topic_v2(topic, cid)
    raise ValueError("getRetry fail, format is wrong")
